pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -319,11 +319,16 @@ private boolean createDictionaryForColumn(ColumnIndexCreationInfo info, SegmentG
       return false;
     }
 
+    // Do not create dictionaries for json or text index columns as they are high-cardinality values almost always
+    if (config.getJsonIndexCreationColumns().contains(column)
+        || config.getTextIndexCreationColumns().contains(column)) {
+      return false;
+    }
+
     // Do not create dictionary if index size with dictionary is going to be larger than index size without dictionary
     // This is done to reduce the cost of dictionary for high cardinality columns
     // Off by default and needs optimizeDictionaryEnabled to be set to true
-    if (config.isOptimizeDictionaryForMetrics() && spec.getFieldType() == FieldType.METRIC && spec.isSingleValueField()
-        && spec.getDataType().isFixedWidth()) {
+    if (config.isOptimizeDictionaryForMetrics() && spec.isSingleValueField() && spec.getDataType().isFixedWidth()) {
```

Comment:
suggest adding a test to validate no dictionary is created for dimension types for high cardinality use cases when this is set to true if possible.